### PR TITLE
Android support via fileDescriptor 

### PR DESCRIPTION
**Pull request by kubaraczkowski**
See base #126 
Since libusb supports the peculiar Android USB handling (usb device must be opened from Java, _after_ acquiring permission from the user to do so) it's useful to enable gousb to do the same.

Checked using a basic Android app with a USB serial device in loopback.

Documentation is not yet done, if the code changes are OK, I'll write the docstrings as well

### CODE REVIEW

File: fakelibusb_test.go [hunk 1]
```diff
@@ -108,6 +112,21 @@ func (f *fakeLibusb) getDevices(*libusbContext) ([]*libusbDevice, error) {
 	return ret, nil
 }
 
+func (f *fakeLibusb) wrapSysDevice(ctx *libusbContext, systemDeviceHandle uintptr) (*libusbDevHandle, error) {
+	if int(systemDeviceHandle) > len(fakeDevices)-1 {
```
Comment (zagrodzki) [inline fakelibusb_test.go:116]: I think if you access an index that is not available as a system device, it should return an error instead

File: fakelibusb_test.go [hunk 2]
```diff
@@ -288,11 +307,12 @@ func (f *fakeLibusb) empty() bool {
 
 func newFakeLibusb() *fakeLibusb {
 	fl := &fakeLibusb{
-		fakeDevices: make(map[*libusbDevice]*fakeDevice),
-		ts:          make(map[*libusbTransfer]*fakeTransfer),
-		submitted:   make(chan *fakeTransfer, 10),
-		handles:     make(map[*libusbDevHandle]*libusbDevice),
-		claims:      make(map[*libusbDevice]map[uint8]bool),
+		fakeDevices:       make(map[*libusbDevice]*fakeDevice),
+		fakeLibUsbDevices: []*libusbDevice{},
```
Comment (zagrodzki) [inline fakelibusb_test.go:311]: this would be necessary only if fakeLibUsbDevices was a map - a zero value of a slice (nil) works just fine as the default value.

File: fakelibusb_test.go [hunk 3]
```diff
@@ -301,7 +321,9 @@ func newFakeLibusb() *fakeLibusb {
 		// The contents of these pointers is never accessed.
 		fd := new(fakeDevice)
 		*fd = d
-		fl.fakeDevices[newDevicePointer()] = fd
+		devPointer := newDevicePointer()
+		fl.fakeDevices[devPointer] = fd
+		fl.fakeLibUsbDevices = append(fl.fakeLibUsbDevices, devPointer)
```
Comment (zagrodzki) [inline fakelibusb_test.go:326]: instead, add a new field to the fakeDevice type, and if fakeDevice.sysDevPtr is set, assign the value in the sys devices map. For testing I think it's fine to assume not all devices would be opened and available through a file descriptor.

File: libusb.go [hunk 4]
```diff
@@ -169,13 +173,37 @@ type libusbIntf interface {
...
+}
 
-func (libusbImpl) init() (*libusbContext, error) {
+func (impl libusbImpl) init() (*libusbContext, error) {
 	var ctx *C.libusb_context
-	if err := fromErrNo(C.libusb_init(&ctx)); err != nil {
+
+	var libusb_options [4]C.struct_libusb_init_option // fixed to 4 - there are maximum 4 options
+	n_options := 0
+	if impl.logLevel != C.LIBUSB_LOG_LEVEL_NONE {
```
Comment (zagrodzki) [inline libusb.go:187]: the logging changes are not strictly related to the core functionality of this PR; split to a separate PR please.

File: libusb.go [hunk 5]
```diff
@@ -169,13 +173,37 @@ type libusbIntf interface {
...
+		libusb_options[n_options].option = C.LIBUSB_OPTION_LOG_LEVEL
+		b := make([]byte, 8)
+		binary.LittleEndian.PutUint64(b, uint64(impl.logLevel))
+		copy(libusb_options[n_options].value[:], b)
+		n_options += 1
+	}
+	if impl.useUsbDK {
```
Comment (zagrodzki) [inline libusb.go:194]: ditto here - I don't think this is needed for Android support? Move to a separate PR

File: usb.c [hunk 6]
```diff
@@ -24,3 +24,4 @@ void gousb_set_debug(libusb_context *ctx, int lvl) {
     libusb_set_debug(ctx, lvl);
 #endif
 }
+
```
Comment (zagrodzki) [inline usb.c:27]: revert

File: usb_test.go [hunk 7]
```diff
@@ -15,7 +15,9 @@
 
 package gousb
 
-import "testing"
+import (
+	"testing"
```
Comment (zagrodzki) [inline usb_test.go:19]: revert this block

File: usb_test.go [hunk 8]
```diff
@@ -94,3 +96,35 @@ func TestOpenDeviceWithVIDPID(t *testing.T) {
...
+		vid, pid ID
+		exists   bool
+	}{
+		{0x9999, 0x0001, true},
+		{0x8888, 0x0002, true},
+		{0x1111, 0x1111, true},
+		{0x7777, 0x0003, false},
+	} {
+		// The fake keeps the order of fakeDevices so that it can be indexed by a fake "fileDescriptor"
+		dev, err := ctx.OpenDeviceWithFileDescriptor(uintptr(ix))
```
Comment (zagrodzki) [inline usb_test.go:115]: this conflates the ordering of the test cases with the file descriptor values, I think it's fragile. See my other comment about a map for sys devices.

File: usb_test.go [hunk 9]
```diff
@@ -94,3 +96,35 @@ func TestOpenDeviceWithVIDPID(t *testing.T) {
...
+		if (dev != nil) != d.exists {
+			t.Errorf("OpenDeviceWithFileDescriptor(%d): device != nil is %v, want %v", ix, dev != nil, d.exists)
+		}
+		if err != nil && d.exists { // It's OK to get an error if device doesn't exist
```
Comment (zagrodzki) [inline usb_test.go:119]: the behavior of OpenDeviceWithFileDescriptor creates a small divergence:
the OpenDeviceWithVIDPID returns nil, nil if the device is not found. Here the function will return an error if the device is not found. I think this is fine though - the Open...VIDPID should have returned an error for not found too.

But this change makes it necessary to change the structure of the test, you cannot use the same structure - I think you don't need to distinguish "exists" vs "not exists", instead you need to distinguish errors vs non-errors.   And I think it may make sense to add a separate test case for a "want error", to simplify the code.

**Review by zagrodzki — CHANGES_REQUESTED** [review]

File: fakelibusb_test.go [hunk 10]
```diff
@@ -87,6 +87,8 @@ type fakeLibusb struct {
 	mu sync.Mutex
 	// fakeDevices has a map of devices and their descriptors.
 	fakeDevices map[*libusbDevice]*fakeDevice
+	// fakeLibUsbDevices keeps the order of devices to be accessd by wrapSysDevice
+	fakeLibUsbDevices []*libusbDevice
```
[the remark this answers is not in the document]
Comment (kubaraczkowski) [reply fakelibusb_test.go:91]: Sure, good point. The FDs I randomly saw were e.g. never 0,1,2 as those are 'special'. Certainly a '0' could be messing up somewhere in a test.

File: libusb.go [hunk 11]
```diff
@@ -169,13 +173,37 @@ type libusbIntf interface {
 }
 
 // libusbImpl is an implementation of libusbIntf using real CGo-wrapped libusb.
-type libusbImpl struct{}
+type libusbImpl struct {
+	logLevel         int
+	disableDiscovery bool
+	useUsbDK         bool
```
[the remark this answers is not in the document]
Comment (kubaraczkowski) [reply libusb.go:179]: OK

File: libusb.go [hunk 12]
```diff
@@ -169,13 +173,37 @@ type libusbIntf interface {
 }
 
 // libusbImpl is an implementation of libusbIntf using real CGo-wrapped libusb.
-type libusbImpl struct{}
+type libusbImpl struct {
+	logLevel         int
+	disableDiscovery bool
```
[the remark this answers is not in the document]
Comment (kubaraczkowski) [reply libusb.go:178]: Yes, good point. I wanted to keep the structure as is in the libusb options, but what you propose makes sense since gousb is supposed to be more 'go-like' rather than being just a libusb wrapper.

File: usb.go [hunk 13]
```diff
@@ -168,6 +179,20 @@ func NewContext() *Context {
 	return newContextWithImpl(libusbImpl{})
 }
 
+type ContextOptions struct {
+	LogLevel               LogLevel
+	DisableDeviceDiscovery bool
+	UseUsbDK               bool
+}
+
+func (o *ContextOptions) NewContext() *Context {
```
[the remark this answers is not in the document]
Comment (kubaraczkowski) [reply usb.go:188]: OK. I must say I didn't see that pattern just yet, but that's just my limited experience!
Certainly having `ContextOptions{}.NewContext()` is silly.

File: fakelibusb_test.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (kubaraczkowski) [reply fakelibusb_test.go:91]: Hm, correction about the 0 - it makes no difference since the FDs still need to be iterated from the list of fakeDevices, unless we'd add a 'fakeFD' entry in there?

File: fakelibusb_test.go (hunk 2 above)
Comment (kubaraczkowski) [reply fakelibusb_test.go:311]: Right. Since that became a map now this gets resolved.

File: fakelibusb_test.go (hunk 3 above)
Comment (kubaraczkowski) [reply fakelibusb_test.go:326]: Right, got it.
The default of uintptr/int will be =0 though, should it be treated as a special value? This would be in case of the wrapSysDevice checking if a file descriptor/sysDevPtr is actually valid.

File: usb_test.go (hunk 9 above)
Comment (kubaraczkowski) [reply usb_test.go:119]: Well, since the fakeDevice can have an uninitialised sysDevPtr (e.g. because the system doesn't support that?) there still is a case where a device exists, but is 'not useful'/'not good'. Perhaps the word 'exists' should rather be 'good' or 'useful' ?
I will split it in two test cases, this will actually make it avoid the wording problem and indeed make the cases simpler.

**Comment by kubaraczkowski** [comment]
Thx @zagrodzki for the comments, please have a look at the update.

File: usb.go [hunk 14]
```diff
@@ -210,6 +227,27 @@ func (c *Context) OpenDevices(opener func(desc *DeviceDesc) bool) ([]*Device, er
...
+	handle, err := c.libusb.wrapSysDevice(c.ctx, fd)
+	if err != nil {
+		return nil, fmt.Errorf("Error opening device from file descriptor: %d, %s", fd, err.Error())
+	}
+
+	dev := c.libusb.getDevice(handle)
+
+	desc, err := c.libusb.getDeviceDesc(dev)
+	if err != nil {
+		return nil, err
```
Comment (zagrodzki) [inline usb.go:240]: if you want to annotate an error somewhere, here it would be appropriate perhaps:
fmt.Errorf("device was opened, but get device descriptor failed: %v")

File: usb_test.go [hunk 15]
```diff
@@ -94,3 +94,45 @@ func TestOpenDeviceWithVIDPID(t *testing.T) {
...
+	}{
+		{0x9999, 0x0001, 78},
+		{0x8888, 0x0002, 94},
+	} {
+		dev, err := ctx.OpenDeviceWithFileDescriptor(d.sysDevPtr)
+		if dev == nil {
+			t.Errorf("OpenDeviceWithFileDescriptor(%d): device == nil for a valid device", d.sysDevPtr)
+		}
+		if err != nil {
+			t.Errorf("OpenDeviceWithFileDescriptor(%d): err != nil for a valid device: %v", d.sysDevPtr, err)
```
Comment (zagrodzki) [inline usb_test.go:115]: Fatalf, and move before the `dev == nil` check.

File: usb_test.go [hunk 16]
```diff
@@ -94,3 +94,45 @@ func TestOpenDeviceWithVIDPID(t *testing.T) {
...
+	for _, d := range []struct {
+		vid, pid  ID
+		sysDevPtr uintptr
+	}{
+		{0x9999, 0x0001, 78},
+		{0x8888, 0x0002, 94},
+	} {
+		dev, err := ctx.OpenDeviceWithFileDescriptor(d.sysDevPtr)
+		if dev == nil {
+			t.Errorf("OpenDeviceWithFileDescriptor(%d): device == nil for a valid device", d.sysDevPtr)
```
Comment (zagrodzki) [inline usb_test.go:112]: Fatalf.

Also add a check to see if the obtained device is the one you asked for (does VID/PID match).

File: usb_test.go [hunk 17]
```diff
@@ -94,3 +94,45 @@ func TestOpenDeviceWithVIDPID(t *testing.T) {
...
+
+	for _, sysDevPtr := range []uintptr{
+		7, // set, but does not exist in the fakeDevices array
+		0, // unset
+	} {
+		dev, err := ctx.OpenDeviceWithFileDescriptor(sysDevPtr)
+		if err == nil {
+			t.Errorf("OpenDeviceWithFileDescriptor(%d): got nil error for invalid device", sysDevPtr)
+		}
+		if dev != nil {
```
Comment (zagrodzki) [inline usb_test.go:133]: drop this check - it's enough to check that error was non-nil. The return value is undefined, it may be non-nil, but it shouldn't be used either way.

File: libusb.go [hunk 18]
```diff
@@ -40,6 +40,7 @@ type libusbDevice C.libusb_device
 type libusbDevHandle C.libusb_device_handle
 type libusbTransfer C.struct_libusb_transfer
 type libusbEndpoint C.struct_libusb_endpoint_descriptor
+type libusbLogLevel C.enum_libusb_log_level
```
Comment (zagrodzki) [inline libusb.go:43]: drop (for later PR)

File: fakelibusb_test.go [hunk 19]
```diff
@@ -87,6 +87,8 @@ type fakeLibusb struct {
 	mu sync.Mutex
 	// fakeDevices has a map of devices and their descriptors.
 	fakeDevices map[*libusbDevice]*fakeDevice
+	// fakeSysDevices keeps the order of devices to be accessd by wrapSysDevice
+	fakeSysDevices map[uintptr]*libusbDevice
```
Comment (zagrodzki) [inline fakelibusb_test.go:91]: actually, perhaps just "sysDevices" - this is already inside the fakeLibusb, so "fake" stutters a bit. "fakeDevices" is ok because the type is called "fakeDevice", although perhaps it should also be renamed "devices"...

File: usb.go [hunk 20]
```diff
@@ -168,6 +168,23 @@ func NewContext() *Context {
 	return newContextWithImpl(libusbImpl{})
```
Comment (zagrodzki) [inline usb.go:168]: would it make sense to replace this with "return ContextOptions{}.New()"? It's now becoming essentially a "context with defaults" anyway.

**Review by zagrodzki — CHANGES_REQUESTED** [review]
also documentation please, I think this is good overall.

File: fakelibusb_test.go (hunk 19 above)
Comment (kubaraczkowski) [reply fakelibusb_test.go:91]: You're the boss, but I'd argue that either both should be changed or neither - it's much more readable to see these two named in the same way. You immediately go "ah, right, these are just some mappings of the fakes"

File: usb.go (hunk 20 above)
Comment (kubaraczkowski) [reply usb.go:168]: I agree - it also matches better the 'new' libusb initializer, since the int `libusb_init(libusb_context  ctx)` is deprecated in favour of always calling it with options.

**Comment by kubaraczkowski** [comment]
Updated! Thx for detailed and clear feedback.

File: fakelibusb_test.go (hunk 19 above)
Comment (zagrodzki) [reply fakelibusb_test.go:91]: I'd go with "both" then :)

File: usb.go [hunk 21]
```diff
@@ -165,7 +165,27 @@ func newContextWithImpl(impl libusbIntf) *Context {
 
 // NewContext returns a new Context instance.
 func NewContext() *Context {
-	return newContextWithImpl(libusbImpl{})
+	return ContextOptions{}.New()
+}
+
+// DeviceDiscovery enables a scan of the available USB devices connected to the system
```
Comment (zagrodzki) [inline usb.go:171]: Add a "Defaults to enabled". Also, full sentences, end with a period.

File: usb.go [hunk 22]
```diff
@@ -165,7 +165,27 @@ func newContextWithImpl(impl libusbIntf) *Context {
 
 // NewContext returns a new Context instance.
```
Comment (zagrodzki) [inline usb.go:166]: ...with default ContextOptions

File: usb.go [hunk 23]
```diff
@@ -165,7 +165,27 @@ func newContextWithImpl(impl libusbIntf) *Context {
...
+
+// DeviceDiscovery enables a scan of the available USB devices connected to the system
+type DeviceDiscovery int
+
+const (
+	EnableDeviceDiscovery = iota
+	DisableDeviceDiscovery
+)
+
+// ContextOptions holds optional flags checked when creating a new Context
```
Comment (zagrodzki) [inline usb.go:179]: "optional flags" is too specific, since we might add more parameters to the context creation that do not easily translate into options in libusb init.

Perhaps:
ContextOptions holds parameters for Context initialization.

File: usb.go [hunk 24]
```diff
@@ -165,7 +165,27 @@ func newContextWithImpl(impl libusbIntf) *Context {
...
+type ContextOptions struct {
+	DeviceDiscovery DeviceDiscovery
```
Comment (zagrodzki) [inline usb.go:181]: Document the field here properly, this is the important part that users will be referencing to understand how to initialize the context.

DeviceDiscovery controls USB device discovery. When set to EnableDeviceDiscovery (default), the connected USB devices will be discovered and enumerated, allowing the use os OpenDevices and OpenWithVIDPID. When set to DisableDeviceDiscovery, OpenDevices will not return any devices. OpenDeviceWithFileDescriptor can be used on some systems to open a device that was otherwise discovered through the operating system.

File: usb.go [hunk 25]
```diff
@@ -165,7 +165,27 @@ func newContextWithImpl(impl libusbIntf) *Context {
...
+}
+
+// New creates a Context, taking into account the optional flags contained in ContextOptions
+func (o ContextOptions) New() *Context {
+	return newContextWithImpl(libusbImpl{
+		discovery: o.DeviceDiscovery,
```
Comment (zagrodzki) [inline usb.go:187]: for your consideration: we could also have discovery be a boolean, and set it to `o.DeviceDiscovery == EnableDeviceDiscovery` here. I'm not sure if it would be much better though, because on one hand it removes a bit of complexity and tight coupling from the libusbImpl, but on the other hand it moves it here and uses two mechanisms instead of one.

File: usb.go [hunk 26]
```diff
@@ -210,6 +230,38 @@ func (c *Context) OpenDevices(opener func(desc *DeviceDesc) bool) ([]*Device, er
 	return ret, reterr
 }
 
+// OpenDeviceWithFileDescriptor takes a (Unix) file descriptor of an opened USB device
+// and wraps the library around it.
+// This is particularly useful when working on Android, where the USB device must be opened
+// by the SDK (Java), giving access to the device through the file descriptor (https://developer.android.com/reference/android/hardware/usb/UsbDeviceConnection#getFileDescriptor()).
```
Comment (zagrodzki) [inline usb.go:236]: format multiline to 80 chars if possible - the URL goes on the separate line

File: usb.go [hunk 27]
```diff
@@ -210,6 +230,38 @@ func (c *Context) OpenDevices(opener func(desc *DeviceDesc) bool) ([]*Device, er
 	return ret, reterr
 }
 
+// OpenDeviceWithFileDescriptor takes a (Unix) file descriptor of an opened USB device
+// and wraps the library around it.
+// This is particularly useful when working on Android, where the USB device must be opened
```
Comment (zagrodzki) [inline usb.go:235]: perhaps "is opened" rather than "must be"? I suspect that Android NDK may offer a similar functionality, it's unlikely the "must be opened in Java SDK" is going to be true forever.

File: usb.go [hunk 28]
```diff
@@ -210,6 +230,38 @@ func (c *Context) OpenDevices(opener func(desc *DeviceDesc) bool) ([]*Device, er
...
+// by the SDK (Java), giving access to the device through the file descriptor (https://developer.android.com/reference/android/hardware/usb/UsbDeviceConnection#getFileDescriptor()).
+//
+// Do note that for this to work the automatic device discovery must be disabled, by calling
```
Comment (zagrodzki) [inline usb.go:238]: not "by calling", but perhaps:

discovery must be disabled at the time when the new Context is created, through the use of ContextOptions.DeviceDiscovery. Example:

File: usb.go [hunk 29]
```diff
@@ -210,6 +230,38 @@ func (c *Context) OpenDevices(opener func(desc *DeviceDesc) bool) ([]*Device, er
...
+//
+// An error is returned in case the file descriptor is not valid.
+func (c *Context) OpenDeviceWithFileDescriptor(fd uintptr) (*Device, error) {
+	handle, err := c.libusb.wrapSysDevice(c.ctx, fd)
+	if err != nil {
+		return nil, err
+	}
+
+	dev := c.libusb.getDevice(handle)
+
```
Comment (zagrodzki) [inline usb.go:251]: drop this empty line

File: usb_test.go [hunk 30]
```diff
@@ -94,3 +94,44 @@ func TestOpenDeviceWithVIDPID(t *testing.T) {
...
+		{0x9999, 0x0001, 78},
+		{0x8888, 0x0002, 94},
+	} {
+		dev, err := ctx.OpenDeviceWithFileDescriptor(d.sysDevPtr)
+		if err != nil {
+			t.Fatalf("OpenDeviceWithFileDescriptor(%d): err != nil for a valid device: %v", d.sysDevPtr, err)
+		}
+		if dev == nil {
+			t.Fatalf("OpenDeviceWithFileDescriptor(%d): device == nil for a valid device", d.sysDevPtr)
+		} else if dev.Desc.Vendor != ID(d.vid) || dev.Desc.Product != ID(d.pid) {
```
Comment (zagrodzki) [inline usb_test.go:116]: remove the "else", this is a new if statement

File: usb_test.go [hunk 31]
```diff
@@ -94,3 +94,44 @@ func TestOpenDeviceWithVIDPID(t *testing.T) {
...
+
+func TestOpenDeviceWithFileDescriptorOnMissingDevice(t *testing.T) {
+	ctx := newContextWithImpl(newFakeLibusb())
+	defer ctx.Close()
+
+	for _, sysDevPtr := range []uintptr{
+		7, // set, but does not exist in the fakeDevices array
+		0, // unset
+	} {
+		_, err := ctx.OpenDeviceWithFileDescriptor(sysDevPtr)
```
Comment (zagrodzki) [inline usb_test.go:131]: if _, err := ctx.Open...; err == nil {

**Review by zagrodzki — CHANGES_REQUESTED** [review]

File: libusb.go [hunk 32]
```diff
@@ -169,13 +171,24 @@ type libusbIntf interface {
...
+type libusbImpl struct {
+	discovery DeviceDiscovery
+}
 
-func (libusbImpl) init() (*libusbContext, error) {
+func (impl libusbImpl) init() (*libusbContext, error) {
 	var ctx *C.libusb_context
-	if err := fromErrNo(C.libusb_init(&ctx)); err != nil {
+
+	var libusb_options [4]C.struct_libusb_init_option // fixed to 4 - there are maximum 4 options
```
Comment (zagrodzki) [inline libusb.go:181]: linter has good points :) libusbOpts perhaps

File: libusb.go [hunk 33]
```diff
@@ -169,13 +171,24 @@ type libusbIntf interface {
...
+	n_options := 0
```
Comment (zagrodzki) [inline libusb.go:182]: numOpts

File: libusb.go [hunk 34]
```diff
@@ -169,13 +171,24 @@ type libusbIntf interface {
...
+	if impl.discovery == DisableDeviceDiscovery {
+		libusb_options[n_options].option = C.LIBUSB_OPTION_NO_DEVICE_DISCOVERY
+		n_options += 1
```
Comment (zagrodzki) [inline libusb.go:185]: numOpts++

**Review by zagrodzki — CHANGES_REQUESTED** [review]

File: fakelibusb_test.go (hunk 19 above)
Comment (kubaraczkowski) [reply fakelibusb_test.go:91]: I understand this as "change both" :) 
Will update

File: usb.go (hunk 23 above)
Comment (kubaraczkowski) [reply usb.go:179]: Good point. I always need to remind myself - gousb is _not_ a libusb wrapper. Not _only_ a wrapper at least.

File: usb.go (hunk 25 above)
Comment (kubaraczkowski) [reply usb.go:187]: Yes, correct, I was indeed thinking on what would be the cleanest way of passing that parameter 'deeper', but decided that converting the enum to a boolean would just make it more confusing.

What about having the `DeviceDiscovery` be a boolean instead of a "boolean enum" ?

File: usb.go (hunk 27 above)
Comment (kubaraczkowski) [reply usb.go:235]: It actually sucks a bit, but unfortunately NDK is not able to open USB devices. Yet? 
TBH I thought that the 'must be' would be a nudge for the users to realise that they will need the Java SDK to acquire permissions and open the device.
But you're correct, it's entirely possible that the NDK limitation will be lifted at some point and this docstring will become invalid.

I will change it to "can be opened"

**Comment by kubaraczkowski** [comment]
Adjusted! Thx for the suggestions.

File: usb.go (hunk 26 above)
Comment (zagrodzki) [reply usb.go:238]: still this

File: usb_test.go [hunk 35]
```diff
@@ -113,7 +113,8 @@ func TestOpenDeviceWithFileDescriptor(t *testing.T) {
 		}
 		if dev == nil {
 			t.Fatalf("OpenDeviceWithFileDescriptor(%d): device == nil for a valid device", d.sysDevPtr)
-		} else if dev.Desc.Vendor != ID(d.vid) || dev.Desc.Product != ID(d.pid) {
+		}
+		if dev.Desc.Vendor != ID(d.vid) || dev.Desc.Product != ID(d.pid) {
 			t.Fatalf("OpenDeviceWithFileDescriptor(%d): device's VID/PID %s/%s don't match expected: %s/%s", d.sysDevPtr, dev.Desc.Vendor, dev.Desc.Product, ID(d.vid), ID(d.pid))
```
Comment (zagrodzki) [inline usb_test.go:118]: this can be an Errorf,  I think

File: usb.go (hunk 25 above)
Comment (zagrodzki) [reply usb.go:189]: DeviceDiscovery as a boolean would default to discovery disabled with zero value. That means all users that want to use `ContextOptions` would need to always include a boilerplate DeviceDiscovery: true to get what should be the default behavior. An enum allows us to interpret zero value as enabled.
Another option is to use DisableDeviceDiscovery as you suggested initially, but that leads to double negatives.

File: usb.go (hunk 24 above)
Comment (zagrodzki) [reply usb.go:183]: what I meant is document the field inside the struct - a user that tries to use ContextOptions will look at the documentation of `ContextOptions` in the first place. So I think this is where it makes sense to add the most verbose documentation. The type itself doesn't need as much of a doc content, although it makes sense to document individual enum values with snippets of the text above.

```
// DeviceDiscovery controls the UBS device discovery and enumeration on Context initialization.
type DeviceDiscovery int

const (
  // EnableDeviceDiscovery means the connected USB devices will be  enumerated 
  // on Context initialization.
  EnableDiscovery DeviceDiscovery = iota
  // DisableDeviceDiscovery means the USB devices are not enumerated and OpenDevices will not
  // return any devices. Without device discovery, OpenDeviceWithFileDescriptor can be used
  // to open devices.
  DisableDiscovery
)
```

**Comment by zagrodzki** [comment]
any updates here?

**Comment by kubaraczkowski** [comment]
Just back from holidays, so it will move in a few days :)

**Comment by kubaraczkowski** [comment]
Sorry for the delay, restarting work after holidays took longer than expected :) 
Something more to be done? Thx for the help!

File: usb_test.go (hunk 16 above)
Comment (zagrodzki) [reply usb_test.go:115]: these two above (line 112 and 115) still need to be Fatalf.

With Fatal(f), the test is immediately interrupted. With Error(f), the test continues, but will report a failure at the end with all recorded errors. The principle is "return as much useful test results from a single run as possible". 

Hence:
* if err is not nil, dev is undefined and no further testing can be done (Fatal)
* if dev is nil, there's no device, testing properties of the device makes no sense (Fatal)
* if dev is not nil and err is nil, then dev may have useful properties. Checking these properties likely doesn't affect whether further testing can be done (Error), unless the code after that check has a hard dependency on the data that the previous check validated.

For future reference: if there's enough distinct check flows that each check forms a chain of dependencies, it may make sense to split individual checks into a subtest (t.Run).

File: usb.go (hunk 24 above)
Comment (zagrodzki) [reply usb.go:183]: this is still open

File: usb.go (hunk 29 above)
Comment (zagrodzki) [reply usb.go:253]: still open

File: usb.go (hunk 26 above)
Comment (zagrodzki) [reply usb.go:238]: still open

File: usb.go [hunk 36]
```diff
@@ -163,9 +163,31 @@ func newContextWithImpl(impl libusbIntf) *Context {
...
-// NewContext returns a new Context instance.
+// NewContext returns a new Context instance with default ContextOptions.
 func NewContext() *Context {
-	return newContextWithImpl(libusbImpl{})
+	return ContextOptions{}.New()
+}
+
+// DeviceDiscovery controls USB device discovery.
+// When set to EnableDeviceDiscovery (default), the connected USB devices will be discovered and enumerated, allowing the use os OpenDevices and OpenWithVIDPID.
+// When set to DisableDeviceDiscovery, OpenDevices will not return any devices. OpenDeviceWithFileDescriptor can be used on some systems to open a device that was otherwise discovered through the operating system.
```
Comment (zagrodzki) [inline usb.go:173]: trim to 80 chars - a combination of long lines and a multi-line comment doesn't render very well in the side-by-side diff view.

**Review by zagrodzki — CHANGES_REQUESTED** [review]
there are still some comments open

File: usb.go (hunk 29 above)
Comment (kubaraczkowski) [reply usb.go:253]: I think it's already in place

File: usb.go (hunk 23 above)
Comment (kubaraczkowski) [reply usb.go]: Anything still missing here?

**Comment by kubaraczkowski** [comment]
Sorry for missing a few points.

Thx for the explanation of Errorf and Fatalf.
I think all comments are now resolved, not 100% sure. I put 👍 on everything I've checked.

**Review by zagrodzki — CHANGES_REQUESTED** [review]
only one minor nitpicky comment left

**Comment by zagrodzki** [comment]
a linux build is failing due to too old libusb version (1.0.25 in the build env). I wonder if we should support the older versions, but I think it would get complicated. I'll just update the minor version at the release time. In the meantime, I need to update the build env, stay tuned.

**Comment by zagrodzki** [comment]
> a linux build is failing due to too old libusb version (1.0.25 in the build env). I wonder if we should support the older versions, but I think it would get complicated. I'll just update the minor version at the release time. In the meantime, I need to update the build env, stay tuned.

Upgraded Linux build env to ubuntu 22.04, rebase and we should be good.

**Comment by kubaraczkowski** [comment]
Any hints on how you would approach keeping it work for ubuntu 20.04 while enabling the additional functionality on 22+ ?
I'm just curious how this could be done, if required.

It seems the feature came out in 2019 and should already be present in 20.04 (1.0.23+), not sure though. 20.04 is still being used, at least EOL is 2025.

**Comment by zagrodzki** [comment]
> Any hints on how you would approach keeping it work for ubuntu 20.04 while enabling the additional functionality on 22+ ? I'm just curious how this could be done, if required.
 
I think I would go with compiler conditionals in the C code, adding a replacement definition of the `libusb_init_context` if the libusb doesn't provide it  (`#if (LIBUSB_API_VERSION < ...)`). And it can inspect the options and if they are non-default, return a "not supported" error.

Alternatively, the libusb version could be inspected from the Go code, so perhaps the C implementation can be trivial (just to appease the compiler, always return a not supported error) and the Go code can conditionally call either old-style init (if old libusb and init called with default options), or fail (old libusb and non-default options), or libusb_init_context if supported.

Both of these would keep working for the existing users, until someone tries to use ContextOptions{} to change some defaults.

> It seems the feature came out in 2019 and should already be present in 20.04 (1.0.23+), not sure though. 20.04 is still being used, at least EOL is 2025.

https://libusb.sourceforge.io/api-1.0/group__libusb__lib.html#gaabdf0bf6be87e12023b7399003c50470 says for `libusb_init_context`:

Since version 1.0.27, `LIBUSB_API_VERSION` >= 0x0100010A

1.0.27 came out Feb 1st this year.

**Review by zagrodzki — APPROVED** [review]

**Comment by zagrodzki** [comment]
thank you for your patience during the process! I'll see if I can get any of the other pending pull requests in before I release a new version, I plan to do that in ~2wks.

**Comment by kubaraczkowski** [comment]
Hey, nice! 

I was looking at only `libusb_wrap_sys_device()` which is the base of android support.
https://libusb.sourceforge.io/api-1.0/group__libusb__dev.html#ga98f783e115ceff4eaf88a60e6439563c

OK, I'm sure somebody using gousb on an older system will complain if needed!

Thanks for the great support!

**Comment by bluetech** [comment]
Regarding @zagrodzki [previous comment](https://github.com/google/gousb/pull/130#issuecomment-2336800368), I'd note that I *do* think it will be a problem in practice once the next version of gousb is released. libusb 1.0.27 is pretty new, and popular targets like e.g. Ubuntu 22.04 LTS have older versions. (I hit it because I tried to use git master to get another unreleased fix).
